op-service/dial/active_l2_provider.go

```diff
@@ -0,0 +1,128 @@
+package dial
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/ethereum/go-ethereum/log"
+)
+
+const DefaultActiveSequencerFollowerCheckDuration = 2 * DefaultDialTimeout
+
+type ethDialer func(ctx context.Context, timeout time.Duration, log log.Logger, url string) (EthClientInterface, error)
+
+type ActiveL2EndpointProvider struct {
+	ActiveL2RollupProvider
+	currentEthClient EthClientInterface
+	ethDialer        ethDialer
+}
+
+func NewActiveL2EndpointProvider(
+	ctx context.Context,
+	ethUrls, rollupUrls []string,
+	checkDuration time.Duration,
+	networkTimeout time.Duration,
+	logger log.Logger,
+	ethDialer ethDialer,
+	rollupDialer rollupDialer,
+) (*ActiveL2EndpointProvider, error) {
+	if len(rollupUrls) == 0 {
+		return nil, errors.New("empty rollup urls list")
+	}
+	if len(ethUrls) != len(rollupUrls) {
+		return nil, errors.New("number of eth urls and rollup urls mismatch")
+	}
+
+	rollupProvider, err := NewActiveL2RollupProvider(ctx, rollupUrls, checkDuration, networkTimeout, logger, rollupDialer)
+	if err != nil {
+		return nil, err
+	}
+	cctx, cancel := context.WithTimeout(ctx, networkTimeout)
+	defer cancel()
+	ethClient, err := ethDialer(cctx, networkTimeout, logger, ethUrls[0])
+	if err != nil {
+		return nil, fmt.Errorf("dialing eth client: %w", err)
+	}
+	return &ActiveL2EndpointProvider{
+		ActiveL2RollupProvider: *rollupProvider,
+		currentEthClient:       ethClient,
+		ethDialer:              ethDialer,
+	}, nil
+}
+
+func (p *ActiveL2EndpointProvider) EthClient(ctx context.Context) (EthClientInterface, error) {
+	p.clientLock.Lock()
+	defer p.clientLock.Unlock()
+	err := p.ensureActiveEndpoint(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	return p.currentEthClient, nil
+}
+
+func (p *ActiveL2EndpointProvider) ensureActiveEndpoint(ctx context.Context) error {
+	if !p.shouldCheck() {
+		return nil
+	}
+
+	if err := p.findActiveEndpoints(ctx); err != nil {
+		return err
+	}
+	p.activeTimeout = time.Now().Add(p.checkDuration)
+	return nil
+}
+
+func (p *ActiveL2EndpointProvider) findActiveEndpoints(ctx context.Context) error {
```

Comment:
This is almost completely duplicating the logic of `ActiveL2RollupProvider.findActiveEndpoints` by the embedded `ActiveL2RollupProvider`. So if we make a change there, we have to carefully apply it here.

I'm wondering if there's a way to reuse `ActiveL2RollupProvider.findActiveEndpoints` by just calling it here, and then on success dialing the `EthClient` for the active index that has been determined. You could then also throw away the duplicated `dialNextSequencer`.

Reply:
Yeah, I was on the fence about duplicating the code vs. "doing the same thing in two places isn't enough to worry about DRY yet". But it's weird that some of the code is duplicated but some uses the embedded logic. I'll find a way to try and make it almost-entirely-embedded as you suggest in the second paragraph there, there's gotta be a way.